pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -190,15 +195,38 @@ func (a *AmbientIndexImpl) All() []*model.AddressInfo {
 	a.mu.RLock()
 	defer a.mu.RUnlock()
 	res := make([]*model.AddressInfo, 0, len(a.byUID)+len(a.serviceByNamespacedHostname))
+
+	addedWEAddr := map[networkAddress]*model.AddressInfo{}
+	allNetworkAddresses := sets.New[networkAddress]()
+
 	for _, wl := range a.byUID {
-		res = append(res, workloadToAddressInfo(wl.Workload))
+		addressInfo := workloadToAddressInfo(wl.Workload)
+		networkAddr := networkAddressFromWorkload(wl)
+		// For WorkloadEntry we need to check whether a workload with similar network/address exists.
+		// WorkloadEntry may have a single address.
+		if strings.Contains(wl.GetUid(), workloadEntryGroupKind) && len(networkAddr) > 0 {
```

Comment:
I get why the approach taken is done, but I don't feel great about this. UID is an opaque string, we shouldn't reference its contents in anyway. I am also a bit concerned about the "lookup and delete from the list"; Delete is super expensive in a slice.

Reply:
UID here can represents more than one stuff, feel not good about its implement

Reply:
I still think we should store the workload entry, since it may become valid/invalid as pods enter and leave the mesh